packages/git-diff/lib/git-diff-view.js

```diff
@@ -1,64 +1,143 @@
-      atom.commands.add(editorElement, 'git-diff:move-to-next-diff', () =>
-        this.moveToNextDiff()
-      ),
-      atom.commands.add(editorElement, 'git-diff:move-to-previous-diff', () =>
-        this.moveToPreviousDiff()
-      ),
-      atom.config.onDidChange('git-diff.showIconsInEditorGutter', () =>
-        this.updateIconDecoration()
-      ),
-      atom.config.onDidChange('editor.showLineNumbers', () =>
-        this.updateIconDecoration()
-      ),
-      editorElement.onDidAttach(() => this.updateIconDecoration()),
-      this.editor.onDidDestroy(() => {
-        this.cancelUpdate();
-        this.removeDecorations();
-        this.subscriptions.dispose();
-      })
+      atom.project.onDidChangePaths(subscribeToRepository)
     );
+  }
+
+  /**
+   * @describe Handles tear down of destructables and subscriptions.
+   *   Does not handle release of memory. This method should only be called
+   *   just before this object is freed, and should only tear down the main
+   *   object components that are guarunteed to exist at all times.
+   */
+  destroy() {
+    // This entire object will be free soon, no need to release here.
+    this.subscriptions.dispose();
+    this.destroyChildren();
+  }
+
+  /**
+   * @describe Destroys this objects children (non-freeing), it's intended
+   *   to be an ease-of use function for maintaing this object. This method
+   *   should only tear down objects that are selectively allocated upon
+   *   repository discovery.
+   *
+   *   Example: this.diffs only exists when we have a repository.
+   */
+  destroyChildren() {
+    if (this._animationId) cancelAnimationFrame(this._animationId);
+
+    if (this.diffs)
+      for (const diff of this.diffs) this.markers.get(diff).destroy();
+  }
+
+  /**
+   * @describe The memory releasing complement function of `destroyChildren`.
+   *   frees the memory allocated at all child object storage locations
+   *   when there is no repository.
+   */
+  releaseChildren() {
+    this.diffs = null;
+    this._repoSubs = null;
+    this._animationId = null;
+  }
+
+  /**
+   * @describe handles all subscriptions based on the repository in focus
+   */
+  async subscribeToRepository() {
+    if (this._repoSubs != null) {
+      this._repoSubs.dispose();
+      this.subscriptions.remove(this._repoSubs);
+    }
+
+    this.repository = await repositoryForPath(this.editor.getPath());
+    if (this.repository != null) {
+      const editorElement = atom.views.getView(this.editor);
+
+      const subscribeToRepository = this.subscribeToRepository.bind(this);
+      const updateIconDecoration = this.updateIconDecoration.bind(this);
+      const scheduleUpdate = this.scheduleUpdate.bind(this);
 
-    this.updateIconDecoration();
-    this.scheduleUpdate();
+      // Every time the repo is changed, the editor needs to be reinitialized.
+      this.subscriptions.add(
+        (this._repoSubs = new CompositeDisposable(
```

Comment:
Please avoid mixing returning and the assignment.

Reply:
Okay, thought it was more readable like that.